### PR TITLE
Feature - connect & pass traces to the jaeger agent using UDP and thrft. (New PR)

### PR DESCRIPTION
**Pull request by saktib**
This is a duplicate PR of https://github.com/open-telemetry/opentelemetry-php/pull/246, as my fork was pretty old from the time I raised the other PR it had lot of conflicts. 

Once this PR will pass we can close the other one and merge this.
 

- Added thrift generator script to pull down jaegertracing/thrift:0.13
- Ran script to generate Jaeger/Thrift library.
- Modified composer to add "apache/thrift": "^0.13.0" dependency for the Jaeger/Thrift library.
- Thrift generation script modified, regenerated Thrifts, added Jaeger PSR4 autoloader in the composer to use the namespace Jaeger inside, other classes.
- Created Jaeger agent exporter(to handle UDP transports), jaeger agent transporter (wrapper above the thrift udp transporter) & modified the thrift udp transporter to make it a little robust.
- Make file updated to accommodate 'make thrift'
- .gitignore updated to exclude Jaeger/Thrift
- Removed php8 from Dockerfile and php.yml to support apache/thrift 0.13.0
- Ran make style,
- Added span converter for generating Jaeger Thrift Span format array to be used in emitBatch.
- Standardized Jaeger Transport class by implementing the new Transport interface.

### CODE REVIEW

File: Jaeger/Codec/CodecUtility.php [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+     *
+     * In case where there is not enough entropy for random_bytes() the generation will use a simpler method.
+     *
+     * @param int $length of bytes
+     * @return string
+     */
+    public static function randomHex(int $length): string
+    {
+        try {
+            return bin2hex(random_bytes($length));
```
Comment (bobstrecansky) [inline Jaeger/Codec/CodecUtility.php:57]: You might want to use `generateTraceId` here

**Review by bobstrecansky — COMMENTED** [review]
This is a great start!   A couple questions to get us rolling.

Are you planning on adding tests and examples with this PR as well?

File: .github/workflows/php.yml [hunk 2]
```diff
@@ -14,7 +14,7 @@ jobs:
     strategy:
       matrix:
         operating-system: [ubuntu-latest]
-        php-versions: ['7.3', '7.4', '8.0']
+        php-versions: ['7.3', '7.4']
```
[the remark this answers is not in the document]
Comment (zsistla) [reply .github/workflows/php.yml:17]: I agree with Bob.  Great work and thanks for putting it up.  Additional tests would be great!  As an example, see how this PR added [exporter](https://github.com/open-telemetry/opentelemetry-php/pull/272/files#diff-06ecd082cc1c55c574b66107b4991900b4f8f3b7ebb347c5887d62faef77d649R1) and [spanconverter](https://github.com/open-telemetry/opentelemetry-php/pull/272/files#diff-5b37a627ea24394698a6909d32e65177b6fbaceb668199aba3f9e7b871101103R1) tests.

**Comment by saktib** [comment]
> Are you planning on adding tests and examples with this PR as well?

I'm planning a second PR for tests

> This is a great start! A couple questions to get us rolling.
> 
> Are you planning on adding tests and examples with this PR as well?

I'm going to add the examples with this PR. However, for tests will create a separate PR.

File: .github/workflows/php.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (saktib) [reply .github/workflows/php.yml:17]: > I agree with Bob. Great work and thanks for putting it up. Additional tests would be great! As an example, see how this PR added [exporter](https://github.com/open-telemetry/opentelemetry-php/pull/272/files#diff-06ecd082cc1c55c574b66107b4991900b4f8f3b7ebb347c5887d62faef77d649R1) and [spanconverter](https://github.com/open-telemetry/opentelemetry-php/pull/272/files#diff-5b37a627ea24394698a6909d32e65177b6fbaceb668199aba3f9e7b871101103R1) tests.

@zsistla , thanks for the PR reference for adding the tests, will follow the same.

File: Jaeger/Codec/CodecUtility.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+     * Thrift package.
+     *
+     * On 32-bit architectures, it falls back to base_convert.
+     *
+     * @return int
+     */
+    public static function getValidI64(int $length) : int
+    {
+        $hex = bin2hex(random_bytes($length));
+        // If we're on a 32-bit architecture, fall back to base_convert.
```
[the remark this answers is not in the document]
Comment (saktib) [reply Jaeger/Codec/CodecUtility.php:35]: As nowadays most of the systems running on 64 bit, will remove the commented code.

File: .github/workflows/php.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (saktib) [reply .github/workflows/php.yml:17]: > Does this PR still not work with 8.0? I'm wondering if it does after some of the underlying libraries might be updated.

We need to wait for this package (which is the required library for UDP transfer) https://packagist.org/packages/apache/thrift#v0.14.1 to be 8.0 compliant.

**Comment by bobstrecansky** [comment]
@saktib - do you have any outstanding things you'd like to do with this PR?

**Comment by bobstrecansky** [comment]
@saktib - there is more discussion about this PR in the slack channel if you'd like to use it to finish up this PR.

**Comment by Grunet** [comment]
@saktib did you happen to still be planning to work on this one? I'm taking initial steps on the very related #296 and was wondering if there was anything I could do to help.

**Comment by saktib** [comment]
> @saktib did you happen to still be planning to work on this one? I'm taking initial steps on the very related #296 and was wondering if there was anything I could do to help.

@Grunet definitely need a helping hand to complete few pending items mentioned below on this PR
> Data type fixes for trace id & span id.
> Unit tests.

Let's talk about this on slack.

File: .github/workflows/php.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Grunet) [reply .github/workflows/php.yml:17]: FWIW the latest version of apache/thrift does appear to call out PHP 8 support https://packagist.org/packages/apache/thrift#v0.15.0

File: Jaeger/Codec/CodecUtility.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
+<?php
```
Comment (Grunet) [inline Jaeger/Codec/CodecUtility.php:1]: How is this file used? I've been unable to figure that out at least

File: Jaeger/Codec/CodecUtility.php (hunk 1 above)
Comment (Grunet) [reply Jaeger/Codec/CodecUtility.php:57]: I haven't been able to figure out what this comment was referring to fwiw

Comment (brettmc) [reply Jaeger/Codec/CodecUtility.php:57]: @Grunet I think to replace the `randomHex` function with `OpenTelemetry\SDK\Trace\RandomIdGeneratoe::generateTraceId()` (although it now has its own `::randomHex()` which is nearly identical to this one, so probably the better choice)

File: Jaeger/Codec/CodecUtility.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (brettmc) [reply Jaeger/Codec/CodecUtility.php:35]: I think that the useful parts of this class could be moved into `SDK\Trace\RandomIdGenerator()`, where similar functionality already/now lives.

File: Jaeger/JaegerTransport.php [hunk 5]
```diff
@@ -0,0 +1,111 @@
+<?php
```
Comment (brettmc) [inline Jaeger/JaegerTransport.php:1]: I think this and related classes should move into `src/Contrib/Jaeger`

**Comment by brettmc** [comment]
Closing, as this PR has been replaced by #545

**Pull request closed without merging**